Add a directory of files used for testing the license list website generating code

goneall opened the pull request.

goneall committed:
Add test files for license generator

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Fix encoding issues

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Sync up test files with current repository

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Update the CECILL text to the text file references rather than a copy/paste of the HTML

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

wking commented on README.md:
```diff
@@ -7,3 +7,8 @@ number as part of the name.
+October 28 2017
+---------------
+
+Added a directory simpleTestForGenerator containing test files for the SPDX License Generator.  These files should be close to the canonical text for the listed license files.
```
I'd rather drop “close to”.  They should *be* the text form (for `.txt` files) of our canonical text.

Also, why `simpleTestForGenerator`?  If the values are the canonical license text, then why not name the directory `canonical`?  The paths should encode what the data *is*, not what we expect it to be used for.

And the `{use-case}/{license}.{extention}` approach inverts the scheme I proposed in #3.  I'd rather keep the license as the first element and use `{license}/license/match/canonical.txt` for these.

goneall commented:
> I'd rather drop “close to”. They should be the text form (for .txt files) of our canonical text.

In the past, there has been some debate as to "what is canonical".  The "close to" was added to acknowledge not everyone believes the text is canonical.  I definitely don't want to be blocked adding text for tests until those debates are resolved.

> Also, why simpleTestForGenerator?

We are still resolving the naming convention for the test repository.  I just picked something simple since I wanted to make the tests available for review without having to wait for the naming conventions to be resolved.

Once we resolve the naming conventions, we can move the files to the appropriate repository naming conventions.

wking commented:
> In the past, there has been some debate as to "what is canonical".

And that makes sense for upstream -> license-list-XML, especially for HPND and similar templates which have no clear upstream.  But *this* repository can ignore "what does upstream think is canonical?" (which can be tough) and focus on "what has the SPDX license list chosen as canonical?" (which is easy).  Does that address your concerns?

> Once we resolve the naming conventions, we can move the files...

Of we're comfortable landing the files, we're approving their naming convention.  That doesn't mean we can't change our minds later.  But I don't see a point in dragging things out.  If the maintainers prefer use-case grouping to license-grouping, I can hold my nose, reroll #3, and we can get it landed.

goneall commented:
@jmanbeck Do you want to send out a summary of the current naming convention proposal?  Perhaps we can close on this and I'll rename my files to the new naming convention.

goneall committed:
Fix CC-BY-ND-2.0 to match the cononical text from creative commons

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall commented:
After discussing the process for submitting tests for the license XML with @jmanbeck , we would like to propose the files for testing the license list XML be included in the license list XML repository itself.  This would make it much easier to submit a new or modified license list XML file along with license text to test against.

We would still keep this repo for testing license scanners and other tools.

Propose a directory structure:

`test/original/[license-id].txt`

The definition of "original" would be the text the license-list-XML is based on.

We could also use "canonical" instead of "original", but I'm trying to avoid debates on whether the text in the license list XML is truly canonical.

In the future, we could add additional subdirectories under test for different types of tests (e.g. license headers, alternative license texts, negative test cases).

wking commented:
On Fri, Dec 15, 2017 at 08:11:40PM +0000, goneall wrote:
> … we would like to propose the files for testing the license list
> XML be included in the license list XML repository itself.  This
> would make it much easier to submit a new or modified license list
> XML file along with license text to test against.

I would rather keep the test files separate, to decouple the content
we expect to match from the templates we use to match it.  For
example, a separate test-data repository like this would not have been
affected by the ongoing license-list → license-list-XML repository
switch.

> We would still keep this repo for testing license scanners and other
> tools.

If we're going to continue to maintain this repository, I really see
no reason to duplicate (some of?) its content in license-list-XML.
For adding new licenses, I propose we add the test data here first
(possibly submitted by legal team or their proxies [1]) after which
the tech folks follow-up with license-list-XML PRs.

Travis testing in license-list-XML (spdx/license-list-XML#503) would
test against a specific version of license-test-files [2], so you
don't have to worry about license-test-files getting out ahead of you
without an explicit bump in the license-list-XML repo.

[1]: https://github.com/spdx/license-test-files/issues/2#issuecomment-331223537
[2]: https://github.com/spdx/license-list-XML/pull/513#issuecomment-351613640

goneall commented:
> I would rather keep the test files separate, to decouple the content
we expect to match from the templates we use to match it.  

I actually want them coupled, in a similar way we would want unit tests to be coupled with the source code under test.

Coupling them also makes it easier to maintain.  

For example, when we implement automated testing of the license list per [License List XML Issue 503](https://github.com/spdx/license-list-XML/issues/503), any additional licenses added will require the addition of a test file.

If we have the test directory in the same repository as the license XML, a submitter would simply add both the license XML and the test file in the same PR.

If we have the test directory in a separate repository, you would need to do the following:

1. Create a PR for adding the license text to the test repository
2. Create a separate PR for the license list XML
3. Somehow get the added license text from the test repository PR pulled into a directory so that the tests would execute correctly.  The only way I know of doing this is to have the PR in the test directory merged, or write some code using the Github API's to try and correlate the PR's

wking commented:
On Fri, Dec 15, 2017 at 12:59:15PM -0800, goneall wrote:
> I actually want them coupled, in a similar way we would want unit
> tests to be coupled with the source code under test.

I'm fine with coupling the test code (although in this case the bulk
of the test code will be separate in spdx/tools).  But I'd rather not
couple the test-data in this case, because I expect different
contribution flows for the test-data, which is fundamentally a
legal-team decision, and the XML templates, which seem closer to the
tech-team.

Decisions about this test data hinge on:

* Should this example text match this identifier?
* Which of several possible matching versions will the SPDX prefer
  under [1]?

and those aren't technical decisions.

Decisions about the XML templates include the format (XML or not),
schema specifics, whether <alt> supports sub-word splitting or only
whole-word splitting, etc.  I expect most of the folks making the
test-data calls do not care about that level of detail.

Having two separate repositories makes it easier for both populations
to make their respective decisions, without getting distracted by the
other's work.

> For example, when we implement automated testing of the license list
> per [License List XML Issue
> 503](https://github.com/spdx/license-list-XML/issues/503), any
> additional licenses added will require the addition of a test file.

I agree with this.

> If we have the test directory in the same repository as the license
> XML, a submitter would simply add both the license XML and the test
> file in the same PR.

That's nice when the submitter has time for it.  But I think the
current backlog is because we have one group of people signing off on
“we want to add ${TEXT} to the license list as ${IDENTIFIER}” and
those folks are generally not the same folks who are submitting the
XML templates.  And even when they are the same person, I don't think
two separate PRs is a high cost.

> If we have the test directory in a separate repository, you would
> need to do the following:
>
> 1. Create a PR for adding the license text to the test repository
> 2. Create a separate PR for the license list XML
> 3. Somehow get the added license text from the test repository PR
>    pulled into a directory so that the tests would execute
>    correctly.  The only way I know of doing this is to have the PR
>    in the test directory merged…

This is what I think we want.  With this repository containing the
authoritative text to match, I don't see why you'd want to go through
the effort of working up template XML before the text had landed here.

Once the PR does land here, the XML PR would just bump the Git commit
ID getting pulled for the XML Travis tests.

[1]: https://spdx.org/licenses/

goneall commented:
> But I'd rather not couple the test-data in this case, because I expect different contribution flows for the test-data

I am thinking about very simple test data - in fact, just a copy/paste of the original text of the license being tested into a simple text file.

In this case, the extra overhead of using 2 repositories is very high compared to a simple copy/paste.  This would be easy for someone who is already submitting an XML file.

We can still maintain a second repo for more sophisticated test data.  

Since I am considering a requirement that the simple test data be added prior to adding a new license XML document, I want to make the process as easy as possible.

Using code development as an analogous workflow, having developers write unit tests and maintain them in the same repository (along with any test data required for the unit tests) is common.  Automating the unit test runs on every push or pull request is also a common practice.  It is also a common practice to have separate system tests in separate repos that are more extensive and maintained by different groups. 

I'm just trying to get the unit tests going since I grow weary finding and fixing the issues when I run the tool for publishing the information on the website.

wking commented:
> Using code development as an analogous workflow, having developers write unit tests and maintain them in the same repository (along with any test data required for the unit tests) is common... It is also a common practice to have separate system tests in separate repos that are more extensive and maintained by different groups.

Agreed, although I'm not aware of cases where all internal tests are formally approved by a separate team (as our legal team will approve these tests).  And I'm also not aware of a project with internal tests that are a strict subset of an external test suite.  In my experience, the internal tests are more extensive.

> I'm just trying to get the unit tests going since I grow weary finding and fixing the issues when I run the tool for publishing the information on the website.

Agreed.  But if you can move that forward based on license-list-XML maintainer consenses, I'd expect you could move it forward here too.  Are the maintainer sets significantly different?

goneall commented:
> Agreed, although I'm not aware of cases where all internal tests are formally approved by a separate team (as our legal team will approve these tests).

Good point - having the original text being approved by legal is probably a good thing.  I could see having a broader set of tests requiring approval by the legal team could be more review work than the legal team would want to take on.

> In my experience, the internal tests are more extensive.

Same here, but just trying to get a start on the tests with something easy to implement.

> Are the maintainer sets significantly different?

@wking you are a significant contributor to both.  I'm just trying to fix things and make the tools work.  There are about 5 or 6 contributors to the license list (caveat - I haven't done any analysis, just based on watching the commit messages).  One of the motivations to moving from a spreadsheet to a git repo of XML files was to open up the contributions to a larger group.  Probably makes sense to open this discussion up to the larger group.  I was thinking of bringing up the automated testing and requirement to add a test file in this Thursday's call.  We could get feedback on the different approaches then.

Based on the discussion so far, I'm changing my opinion a bit.  I'm now thinking we maintain 2 set of files which are used for test.  One set of files is the "original" text we can use to verify the XML files do indeed match the text it is intended to represent.  The other set are more extensive tests.  There may (or may not) be overlap, but I don't see much issue since the "original text" really should change.

goneall committed:
Add new test case for added license

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

wking commented:
> ... I don't see much issue since the "original text" really should change.

It sounds like you meant "should *not* change".  But see:

* spdx/license-list-XML#497, which adds a title to our canonical text.
* spdx/license-list-XML#516, which significantly rewords our canonical text

And I think it would have been presumptive to land [the optional "/or"][1] without legal-team approval.  Implementing templates for these is tech-team stuff, but approving themas canonical changes or accepted matches seems like legal-team stuff to me.

[1]: https://github.com/spdx/license-list-XML/pull/423#issuecomment-318797334

goneall committed:
Add CLDA-Sharing-1.0 to tests

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Add test files for BSD-1-Clause and EPL-2.0

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Fix issues identified in license list XML PR #513

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Add license tests for the new licenses added

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

goneall committed:
Add missing copyright to test file

Signed-off-by: Gary O'Neall <gary@sourceauditor.com>

wking committed:
simpleTestForGenerator/BitTorrent-1.0: Fetch from Internet Archive

Generated with:

  $ curl -s https://web.archive.org/web/20050209224501/http://www.bittorrent.com/license |
  >   sed -z 's|.*<pre>[[:space:]]*||;s|[[:space:]]*</pre>.*|\n|' |
  >   sed 's/[[:space:]]*$//' >simpleTestForGenerator/BitTorrent-1.0.txt

goneall committed:
Merge pull request #1 from wking/fix-bittorrent-1.0

simpleTestForGenerator/BitTorrent-1.0: Fetch from Internet Archive

wking committed:
simpleTestForGenerator/Apache-1.0: Add missing closing quote

Catching up with spdx/license-list-XML@558562e6 (Add missing close
quote to Apache-1.0.xml, 2017-12-30, spdx/license-list-XML#588).

goneall committed:
Merge pull request #2 from wking/apache-1.0-missing-quote

simpleTestForGenerator/Apache-1.0: Add missing closing quote
